Math-Interactions-Overview.md

```diff
@@ -0,0 +1,49 @@
+
+For backend validations of the data stored for these interactions, we have a custom expression parser in `core/domain/expression_parser.py`. **Note that these need to be in sync with the behavior of nerdamer's validations.**
+
+## FAQs
+- **How is the data stored for these interactions validated?**
+  - There are 2 layers of validation for all three math interactions.
+    - **Frontend validations:** This is a set of validations that happen in real-time as soon as the learner/creator enters an expression into guppy. On any change to the guppy input, the input is passed to validation handlers that internally use nerdamer to run syntactic as well as logical checks on the given input. Any errors caught at this stage are displayed to the user in real-time and the submit button is disabled until they are fixed. One caveat to note here is that the checks and display of errors happen when the user clicks outside the input box and not on single-letter changes. This is intentionally done since the latter is an annoying experience. To achieve this, the validations are run when the currently active guppy instance goes out of focus, implying that the user is done typing and validations can be run. Note that these validations run anywhere a guppy instance is present so they will be run on both creator's and learner's view. **These are handled by the `core/templates/services/math-interaction-service.ts`**
+    - **Backend validations:** After the first level of validations from the frontend pass, the creator's input needs to be stored in the datastore. Before this happens, we also run some backend validations on the input as a backup check for any data that gets stored. This is done to ensure that scenarios such as breakages in the frontend, direct backend calls, one-off/audit backend jobs, etc. are all handled. **These are handled by the `core/domain/expression_parser.py`**
+  - Other than this there are obviously some exploration level validations for all these interactions that check exploration level validity to avoid scenarios like redundant rule input, impossible-to-reach state, etc. **These are handled by the individual validation services present in each interaction's directory.**
+
+## Glossary
+- [Guppy](https://guppy.js.org/site/): Opensource tool that provides an input widget for the math interactions. Supports WYSIWYG and provides a lot of APIs that can be used to build customizations on top of the basic tools.
+- [Nerdamer](https://nerdamer.com/): Opensource javascript library for handling (validating, operating, etc) the math expressions on the frontend.
+
+## References
+- [Original proposal](https://drive.google.com/file/d/1vB3vxvBUEsYivUgTiZrmUYhmReCfQPQE/view)
+
+## Open Issues
```

Comment:
Can you add this information to the issue descriptions/threads directly? Otherwise, people working on them won't know that they should find it here.

Reply:
Done.

Reply:
Thanks for adding the context to the issues!

Looking at this again, I suggest dropping this "Open Issues" section and, instead, in each issue mentioned here, linking developers to this wiki page so that they can get the necessary context to tackle the issue. That would make more sense in the "developer's user flow". Would that be OK?

Reply:
Yup, that makes sense. Removed it.